<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>

<body>
    <script>
        /* 
            then方法的返回值： 
                - then一定返回一个promise对象，但是根据情况不同，返回的结果不同：
                - then中是一个回调函数，如果回调函数返回一个值，那么then返回的Promise对象就是成功状态，并且将回调函数的返回值作为promise对象的值 
                - then中是一个回调函数，如果回调函数没有返回值，那么then返回的Promise对象就是成功状态，并且promise对象的值是undefined
                
                - then中是一个回调函数，如果回调函数返回一个成功状态的promise对象，那么then返回的Promise对象也是成功状态，并且值是回调函数返回成功的promise对象的值
                - then中是一个回调函数，如果回调函数返回一个失败状态的promise对象，那么then返回的Promise对象也是失败状态，并且值是回到函数返回失败的promise对象的值
                - then中是一个回调函数，如果回调函数返回一个未定义状态（pending）的promise对象，那么then的返回值状态也是未定义的

                - 如果then中的回调函数抛出一个错误，那么then返回的Promise对象将会成为失败状态，并且将错误作为失败状态的promise的值

                - then中不传递回调函数，则会发生值穿透

        */

        const p1 = new Promise((resolve, reject) => {
            console.log("开始请求a。。。。");

            setTimeout(() => {
                try {
                    //模拟抛出错误（后边代码不再执行直接进入catch）
                    // throw new Error("a数据请求失败")

                    //a请求成功
                    console.log("a数据请求成功");
                    //模拟请求回来的a数据
                    const dataA = [1, 2, 3, 4];
                    //调用成功回调
                    resolve(dataA);
                } catch (e) {
                    console.log(e);
                    //错误处理
                    reject(e);
                }
            }, 1000)

        })

        console.log(p1);

        //调用then方法
        /* const thenRes1 = p1.then(() => {
            return "hello world";
        })
        console.log("thenRes1", thenRes1); */


        /* const thenRes2 = p1.then(() => {
            //没有return
        })
        console.log("thenRes2", thenRes2); */

        /* const thenRes3 = p1.then(() => {
            return new Promise((resolve, reject) => {
                resolve("then中promise成功")
            })
        })
        console.log("thenRes3", thenRes3); */


        /* const thenRes4 = p1.then(() => {
            return new Promise((resolve, reject) => {
                reject("then中promise失败")
            })
        })
        console.log("thenRes4", thenRes4); */

        /* const thenRes5 = p1.then(() => {
            throw new Error("then中报错")
            return new Promise((resolve, reject) => {

                resolve("then中promise成功")
            })
        })
        console.log("thenRes5", thenRes5); */
    </script>
</body>

</html>